public/app/features/plugins/admin/state/actions.ts

```diff
@@ -18,16 +19,74 @@ import { STATE_PREFIX } from '../constants';
 import { mapLocalToCatalog, mergeLocalsAndRemotes, updatePanels } from '../helpers';
 import { CatalogPlugin, RemotePlugin, LocalPlugin } from '../types';
 
+// Fetches
 export const fetchAll = createAsyncThunk(`${STATE_PREFIX}/fetchAll`, async (_, thunkApi) => {
```

Comment:
I think we can make this call a bit simpler by doing the following:

```ts
    const pluginErrors = from(getPluginErrors());
    const localPlugins = from(getLocalPlugins());
    
    // Will create an observable that will emit empty array as first value
    // It will also emit empty array if an error occur.
    const remotePlugins = from(getRemotePlugins()).pipe(
      startWith([]),
      catchError(() => {
        thunkApi.dispatch({ type: `${STATE_PREFIX}/fetchRemote/rejected` });
        return from([]);
      })
    );


    // By using combine latest the subscribe function will emit when all observables
    // has emitted at least one value. After that it will emit the latest value from each observable
   // every time any of the observables will emit a new value until all of them completes.
    combineLatest([localPlugins, remotePlugins, pluginErrors]).subscribe(([local, remote, errors]) => {
      console.log(Date.now(), { local, remote, errors });

      if (local.length > 0) {
        // is it OK if this will be dispatched multiple times?
        thunkApi.dispatch({ type: `${STATE_PREFIX}/fetchLocal/fulfilled` });
      }

      if (remote.length > 0) {
        // is it OK if this will be dispatched multiple times?
        thunkApi.dispatch({ type: `${STATE_PREFIX}/fetchRemote/fulfilled` });
      }

      thunkApi.dispatch(addPlugins(mergeLocalsAndRemotes(local, remote, errors)));
    });
```

Also, since we are handling an observable here we don't need to create an async thunk but can instead create a normal thunk.

Reply:
Nice, I really like the usage of `combineLatest()`!

I think one of the issues I could be, that the `remotePlugins` is going to be slower than the local queries, and as it is starting with an empty array there will always be a "flickering", even if the remote call is relatively fast. (Maybe it's not that visible though, we would need to test).

Maybe to solve that we could keep a `timeout()` in the `remotePlugins` observable, somehow like this?
```typescript
const remotePlugins = from(getRemotePlugins()).pipe(
      timeout({
         each: 500,
         with: () => {
            // Subscribe to `remotePlugins` and dispatch an action separately? 
            // Or could we do something wiser? 
              
            return from([]);
         }
      }),
      catchError(() => {
        thunkApi.dispatch({ type: `${STATE_PREFIX}/fetchRemote/rejected` });
        return from([]);
      })
    );
```

I was also thinking that we probably wouldn't need to dispatch the `fetchLocal` and `fetchRemote` actions, but just dispatching the `addPlugins` every time (for using the previous two we would need to merge them with the already existing local/remote counterparts in the store)?
```typescript
// By using combine latest the subscribe function will emit when all observables
// has emitted at least one value. After that it will emit the latest value from each observable
// every time any of the observables will emit a new value until all of them completes.
combineLatest([localPlugins, remotePlugins, pluginErrors]).subscribe(([local, remote, errors]) => {
      thunkApi.dispatch(addPlugins(mergeLocalsAndRemotes(local, remote, errors)));
});
```

Reply:
Yeah, I think that makes sense! It still feels a bit easier to understand the flow with these changes 🙌🏻

Reply:
hmm, there might be one thing with the timeout. It might complete the underlying observable after the timeout.

Reply:
Hmm, yea, maybe that's what happens indeed 🤔